lib/praxis/mapper/resource.rb

```diff
@@ -79,6 +89,130 @@ def self.define_model_accessors
+              end
+              orig_call = proc { |**kw| super(**kw)}
+              result = if calls[:around].presence
+                # TODO: This can be a nested loop of sends, without procs...?
+                calls[:around].inject(orig_call) do |inner, target|
+                  proc { |**kw| send(target, **kw, &inner) }
+                end.call(**kwargs)
+              else
+                super(**kwargs)
+              end
+              calls[:after]&.each do |target|
+                target.is_a?(Symbol) ? send(target, **kwargs) : self.instance_exec(**kwargs, &target)
+              end
+              result
+            end
+          end
+        end
+        prepend themodule
+      end
+
+      def self.intercept_callbacks_for_class(method, calls)
+        has_args = method(method).parameters.any? { |(type, _)| [:req, :opt, :rest].include?(type) }
+        themodule = if has_args
+          Module.new do
+            # Setup the method to take both args and  kwargs
+            define_method(method) do |*args, **kwargs|
+              calls[:before]&.each do |target|
+                target.is_a?(Symbol) ? send(target, *args, **kwargs) : self.instance_exec(*args, **kwargs, &target)
+              end
+              orig_call = proc { |*a, **kw| super(*a, **kw)}
+              result = if calls[:around].presence
+                # TODO: This can be a nested loop of sends, without procs...?
+                calls[:around].inject(orig_call) do |inner, target|
+                  proc { |*a, **kw| send(target, *a, **kw, &inner) }
+                end.call(*args, **kwargs)
+              else
+                super(*args, **kwargs)
+              end
+              calls[:after]&.each do |target|
+                target.is_a?(Symbol) ? send(target, *args, **kwargs) : self.instance_exec(*args, **kwargs, &target)
+              end
+              result
+            end
+          end
+        else
+          Module.new do
+            # Setup the method to only take kwargs
+            define_method(method) do |**kwargs|
+              calls[:before]&.each do |target|
+                target.is_a?(Symbol) ? send(target, **kwargs) : self.instance_exec(**kwargs, &target)
+              end
+              orig_call = proc { |**kw| super(**kw)}
+              result = if calls[:around].presence
+                # TODO: This can be a nested loop of sends, without procs...?
+                calls[:around].inject(orig_call) do |inner, target|
+                  proc { |**kw| send(target, **kw, &inner) }
+                end.call(**kwargs)
+              else
+                super(**kwargs)
+              end
+              calls[:after]&.each do |target|
+                target.is_a?(Symbol) ? send(target, **kwargs) : self.instance_exec(**kwargs, &target)
+              end
+              result
+            end
+          end
+        end
+        singleton_class.send(:prepend, themodule)
+      end
```

Comment:
Seems like the `intercept_callbacks_for_x` methods are basically identical except for how they use `themodule`. Worth DRYing the module construction out into its own method?

Reply:
Yeah, was still not fully DRYed yet. It is now. :)